systems/trajectory_optimization/dircon/dircon.cc

```diff
@@ -0,0 +1,843 @@
+    //
+    double min_dt = mode.min_T() / (mode.num_knotpoints() - 1);
+    double max_dt = mode.max_T() / (mode.num_knotpoints() - 1);
+    AddBoundingBoxConstraint(min_dt, max_dt, timestep(mode_start_[i_mode]));
+    for (int j = 0; j < mode.num_knotpoints() - 2; j++) {
+      // all timesteps must be equal
+      AddLinearConstraint(timestep(mode_start_[i_mode] + j) ==
+                          timestep(mode_start_[i_mode] + j + 1));
+    }
+
+    //
+    // Create new decision variables
+    //
+    force_vars_.push_back(NewContinuousVariables(
+        mode.evaluators().count_full() * mode.num_knotpoints(),
+        "lambda[" + std::to_string(i_mode) + "]"));
+    collocation_force_vars_.push_back(NewContinuousVariables(
+        mode.evaluators().count_full() * (mode.num_knotpoints() - 1),
+        "lambda_c[" + std::to_string(i_mode) + "]"));
+    collocation_slack_vars_.push_back(NewContinuousVariables(
+        mode.evaluators().count_full() * (mode.num_knotpoints() - 1),
+        "gamma[" + std::to_string(i_mode) + "]"));
+
+    // quaternion_slack_vars_ (slack variables used to scale quaternion norm to
+    // 1 in the dynamic constraints)
+    int num_quat = multibody::QuaternionStartIndices(plant_).size();
+    quaternion_slack_vars_.push_back(
+        NewContinuousVariables(num_quat * (mode.num_knotpoints() - 1),
+                               "quat_slack[" + std::to_string(i_mode) + "]"));
+
+    // Bound quaternion slack variables to avoid false full rotations
+    double slack_bound = 1;
+    AddBoundingBoxConstraint(-slack_bound, slack_bound,
+                             quaternion_slack_vars_.at(i_mode));
+
+    //  Post-impact variables. Note: impulse variables declared below.
+    if (i_mode > 0) {
+      v_post_impact_vars_.push_back(NewContinuousVariables(
+          plant_.num_velocities(), "v_p[" + std::to_string(i_mode) + "]"));
+    }
+
+    // Constraint offset variables for relative constraints
+    offset_vars_.push_back(
+        NewContinuousVariables(mode.num_relative_constraints(),
+                               "rel_offset[" + std::to_string(i_mode) + "]"));
+
+    //
+    // Create context elements for knot points
+    //
+    for (int j = 0; j < mode.num_knotpoints(); j++) {
+      contexts_[i_mode].push_back(std::move(plant_.CreateDefaultContext()));
+    }
+
+    // We only need an impact if the post-impact constraint set contains
+    // anything not already in the pre-impact constraint set.
+    bool is_impact = false;
+    if (i_mode > 0) {
+      auto prev_evaluators = get_mode(i_mode - 1).evaluators().get_evaluators();
+      for (auto e : mode.evaluators().get_evaluators()) {
+        if (std::find(prev_evaluators.begin(), prev_evaluators.end(), e) ==
+            prev_evaluators.end()) {
+          is_impact = true;
+          break;
+        }
+      }
+    }
+
+    //
+    // Create and add collocation constraints
+    //
+    int cache_size = 1000;
+    cache_.push_back(
```

Comment:
Have you tested different cache sizes? Does 1000 seem to be the sweet spot? If so, maybe make it a constant? If not, add an option to adjust it.

Reply:
I'll set it intelligently. A safe number is equal to the number of decision variables (plus 1) for numerical gradients.